src/main/java/net/sf/mzmine/modules/peaklistmethods/filtering/duplicatefilter/DuplicateFilterTask.java

```diff
@@ -137,92 +139,175 @@ public void run() {
-            // Compare rt
-            final boolean sameRT = rtTolerance.getToleranceRange(firstRow.getAverageRT())
-                .contains(secondRow.getAverageRT());
+            boolean sameMZRT = filterByAvgRTMZ ? // average or single feature
+                checkSameAverageRTMZ(firstRow, secondRow, mzTolerance, rtTolerance)
+                : checkSameSingleFeatureRTMZ(rawFiles, firstRow, secondRow, mzTolerance,
+                    rtTolerance);
 
             // Duplicate peaks?
-            if (sameID && sameMZ && sameRT) {
-
+            if (sameID && sameMZRT) {
+              // create consensus row in new filter
+              if (!mode.equals(FilterMode.OLD_AVERAGE)) {
+                // copy all detected features of row2 into row1
+                // to exchange gap-filled against detected features
+                createConsensusRow(rawFiles, firstRow, secondRow);
+              }
+              // second row deleted
+              n++;
               peakListRows[secondRowIndex] = null;
             }
           }
         }
+        // add to new list
+        newPeakList.addRow(firstRow);
       }
-
       processedRows++;
     }
 
-    // Create the new peak list.
-    final PeakList newPeakList =
-        new SimplePeakList(origPeakList + " " + suffix, origPeakList.getRawDataFiles());
-
-    // Add all remaining rows to a new peak list.
-    for (int i = 0; !isCanceled() && i < rowCount; i++) {
-
-      final PeakListRow row = peakListRows[i];
-
-      if (row != null) {
-
-        // Copy the peak list row.
-        final PeakListRow newRow = new SimplePeakListRow(row.getID());
-        PeakUtils.copyPeakListRowProperties(row, newRow);
-
-        // Copy the peaks.
-        for (final Feature peak : row.getPeaks()) {
-
-          final Feature newPeak = new SimpleFeature(peak);
-          PeakUtils.copyPeakProperties(peak, newPeak);
-          newRow.addPeak(peak.getDataFile(), newPeak);
-        }
-
-        newPeakList.addRow(newRow);
-      }
-    }
-
+    // finalize
     if (!isCanceled()) {
-
       // Load previous applied methods.
       for (final PeakListAppliedMethod method : origPeakList.getAppliedMethods()) {
-
         newPeakList.addDescriptionOfAppliedTask(method);
       }
 
       // Add task description to peakList
       newPeakList.addDescriptionOfAppliedTask(
           new SimplePeakListAppliedMethod("Duplicate peak list rows filter", parameters));
+      LOG.info("Removed " + n + " duplicate rows");
     }
 
     return newPeakList;
   }
+
+  /**
+   * Turns firstRow to consensus row. With all features with highest FeatureStatus:
+   * DETECTED>ESTIMATED>UNKNOWN Or the highest feature when comparing two ESTIMATED features
+   * 
+   * @param rawFiles
+   * @param firstRow
+   * @param secondRow
+   */
+  private void createConsensusRow(RawDataFile[] rawFiles, PeakListRow firstRow,
+      PeakListRow secondRow) {
+    for (RawDataFile raw : rawFiles) {
+      Feature f2 = secondRow.getPeak(raw);

```

Comment:
The code looks a lot better when it is in a separate method, but still this block full of if() statements and brackets is a torture to read. 

I think you can make it more human-readable: 
- change `if (f2 != null) {` to `if (f2 == null) return;`, i.e. exit the method quickly rather than branching the code
- use a switch statement for `f2.getFeatureStatus()`, which will make the code flow more obvious